segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -167,6 +172,56 @@ public void setupAppend(SetupAppend setupAppend) {
                 });
     }
 
+    @VisibleForTesting
+    CompletableFuture<Void> setupTokenExpiryTask(SetupAppend setupAppend) {
+        String segment = setupAppend.getSegment();
+        UUID writerId = setupAppend.getWriterId();
+        long requestId = setupAppend.getRequestId();
+
+        Duration expiryDuration = this.durationToExpiry(setupAppend);
+
+        if (expiryDuration == null) {
+            return CompletableFuture.completedFuture(null);
+        } else {
+            return Futures.delayedTask(() -> {
+                if (isSetupAppendCompleted(segment, writerId)) {
+                    try {
+                        log.trace("Informing writer {} that sent request {}, about token expiry for segment {}",
+                                writerId, requestId, segment);
+                        if (!connection.isClosed()) {
+                            connection.close();
+                        }
+                    } catch (RuntimeException e) {
+                        // Log and ignore
+                        log.warn("Unable to inform writer {} that sent request {}, about token expiry for segment {}",
+                                writerId, requestId, segment);
+                    }
+                }
+                return null;
+            }, expiryDuration, this.tokenExpiryHandlerExecutor);
+        }
+    }
+
+    @VisibleForTesting
+    Duration durationToExpiry(SetupAppend setupAppend) {
+        final Duration duration = JwtUtils.durationToExpiry(setupAppend.getDelegationToken());
+
+        // Note that duration can be null, say if token had no expiry set (for internal communications).
+
+        if (duration != null && duration.isNegative()) {
```

Comment:
Would the `TokenVerifierImpl#verifyToken` verify if the duration is negative? If not, we could move the check there to ensure expired tokens are not allowed.

Reply:
Yes, that check is for finding out if the token has already expired.

Reply:
TokenVerifierImpl#verifyToken is already checking for TokenExpired. Hence this check seems redundant.

Reply:
I've now removed it.